Source/Core/VideoCommon/LightingShaderGen.cpp

```diff
@@ -11,69 +11,97 @@
 #include "VideoCommon/XFMemory.h"
 
 static void GenerateLightShader(ShaderCode& object, const LightingUidData& uid_data, int index,
-                                int litchan_index, bool alpha)
+                                AttenuationFunc attn_func, DiffuseFunc diffuse_func, bool alpha)
 {
+  object.Write("  {{ // {} light {}\n", alpha ? "Alpha" : "Color", index);
   const char* swizzle = alpha ? "a" : "rgb";
   const char* swizzle_components = (alpha) ? "" : "3";
 
-  const auto attnfunc =
-      static_cast<AttenuationFunc>((uid_data.attnfunc >> (2 * litchan_index)) & 0x3);
-  const auto diffusefunc =
-      static_cast<DiffuseFunc>((uid_data.diffusefunc >> (2 * litchan_index)) & 0x3);
+  // Create a normalized vector pointing from the light to the current position.  We manualy
+  // normalize instead of using normalize() because the raw distance is needed for spot lights.
+  object.Write("    float3 ldir = " LIGHT_POS ".xyz - pos.xyz;\n", LIGHT_POS_PARAMS(index));
```

Comment:
As I've worked on passing lighting information to custom shaders, I'm curious why we calculate lighting direction instead of just using the `LIGHT_DIR`?

Reply:
The `dir` field (which `LIGHT_DIR` uses) is only valid for specular lights; it contains garbage data (see https://github.com/dolphin-emu/dolphin/pull/10475#issuecomment-1051287750) for spot lights.

Instead, for spot lights, the light is placed in 3D space sufficiently far away from everything that it's effectively only in one direction, and the vector from the light's position to the current vertex is what's used. This is *also* done for specular lights, but they also compare it with the half angle vector for the light to determine if there should be a highlight (at least I think; it's been a little while since I looked into this).

I think the big issue here is that the names for the various light fields are incredibly misleading and confusing. But I'm not sure how much can be done to fix that.

Reply:
> The dir field (which LIGHT_DIR uses) is only valid for specular lights

Specular == Point light right?

So the `dir` field is special for specular lights and includes a half-angle calculation?

> it contains garbage data for spot lights.

That is surprising.  Shouldn't we not be using `dir` for spot lights then?

> Instead, for spot lights, the light is placed in 3D space sufficiently far away from everything that it's effectively only in one direction

Strange, that doesn't sound much different than the direction light.  I do see it has falloff though, do most games just not use it that way?

> I think the big issue here is that the names for the various light fields are incredibly misleading and confusing.

What would you name them to be clearer?

Reply:
I think I must have confused myself a lot. I know it's garbage data in some cases, but I think they set the cosine attentuation up so that even though `dir` is being used it ends up not mattering (as long as `NaN` doesn't break things), or something like that. It's been a little while.

I believe most games don't use the distance falloff (nor do they use `AttenuationFunc::Dir` I think?) but it is present.

Probably the biggest source of confusion for the names is that there's a union with 2 names in it in XFMemory.

Reply:
Thanks @Pokechu22 .  Sorry for hijacking your PR for my questions but I appreciate the details.